### PR TITLE
[12.0][ADD] account_payment_netting

### PR DESCRIPTION
**Pull request by kittiu**
This module allow net payment on AR/AP invoice from the same business partner.

**NOTE**: This module is influenced by account_netting,
but make it more user friendly when netting invoices.
While account netting require user to select manually the journal items to do netting
(which create netting journal entry), this module has a new menu "Invoices to netting"
allowing user to select both customer/supplier invoice to register payment.

### CODE REVIEW

**Comment by kittiu** [comment]
This module is ready for review. Thanks!

**Comment by kittiu** [comment]
Changed development_status to Beta

**Comment by kittiu** [comment]
@OCA/accounting-maintainers Do you think this module should be in account-payment repo?

**Comment by kittiu** [comment]
@rousseldenis :)

File: account_payment_netting/views/account_invoice_view.xml
Comment (yostashiro) [inline account_payment_netting/views/account_invoice_view.xml:36]: I'd personally put the menu items next to "Customers > Payments" and "Vendors > Payments" (I suppose having the same view in two separate places is OK for this case).

One concern of having the menu item under `account.menu_finance_entries` is that some users may be confused for not enabling "Show Full Accounting Features".

File: account_payment_netting/views/account_invoice_view.xml
Comment (yostashiro) [inline account_payment_netting/views/account_invoice_view.xml]: "Invoices **for** Netting" is better?

**Review by yostashiro — COMMENTED** [review]
Haven't checked the logic in detail but the module seems to work nicely.  @kittiu I think you forgot to run `oca-gen-addon-readme`.

File: account_payment_netting/views/account_invoice_view.xml
Comment (kittiu) [reply account_payment_netting/views/account_invoice_view.xml:36]: Thank you @yostashiro I think oca-gen-addon-readme is not mandatory, as long as we have readme folder it will be auto generated once it is merged (I am yet to see too).
Yes, we intended for only full accounting user to use it at the moment, I think it is not much generic feature (we only have 1 customer request to use it).

File: account_payment_netting/views/account_invoice_view.xml
Comment (kittiu) [reply account_payment_netting/views/account_invoice_view.xml]: OK, done.

**Comment by kittiu** [comment]
This one is ready to be reviewed/merged.

File: account_payment_netting/models/account_invoice.py
Comment (rousseldenis) [inline account_payment_netting/models/account_invoice.py]: @kittiu I would have split this method in some little ones to easier reading and overriding.

File: account_payment_netting/models/account_invoice.py
Comment (rousseldenis) [inline account_payment_netting/models/account_invoice.py]: This could have been put in a compute field instead.

File: account_payment_netting/models/account_invoice.py
Comment (rousseldenis) [inline account_payment_netting/models/account_invoice.py:29]: Maybe a _get_netting_groups() method

File: account_payment_netting/models/account_payment.py
Comment (rousseldenis) [inline account_payment_netting/models/account_payment.py:97]: (I'm not really aware of all accounting stuffs) Is it sufficent for refunds cases ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
LGTM. Some comments about the register_payment function essentially.

**Comment by kittiu** [comment]
@rousseldenis all commented fixed. Thank you for your review :)

File: account_payment_netting/models/account_payment.py
Comment (kittiu) [reply account_payment_netting/models/account_payment.py:97]: Sorry, I just test it and it is not yet able to pay net amount of (customer invoice - customer refund). I think it is good idea to also the case you mentioned. I think I should add 2 extra menus
1. Invoicing > Customers > Invoices for Netting : show only out_invoice, out_refund
2. Invoicing > Vendors > Invoices for Netting : show only in_invoice, in_refund.

Note: I think this will be in good use, but also wonder why Odoo do not allow paying the net of Invoice and Refund from the first place...

Comment (kittiu) [reply account_payment_netting/models/account_payment.py:97]: @rousseldenis I found that the Oodo standard intentionally do not allow invoice and refund together
https://github.com/odoo/odoo/blob/12.0/addons/account/models/account_payment.py#L90
If I just remove this part of logic (and just create the menu that show invoice/payment together in list view), it will just work.

My opinion is, this should be a separated addon, i..e, account_payment_on_invoice_and_refund (still not a good name :p).

And the account_payment_netting is already completed by itself.

WDYT?

Comment (kittiu) [reply account_payment_netting/models/account_payment.py:97]: I think the discussion here is related to this question https://github.com/OCA/account-invoicing/issues/551

Comment (kittiu) [reply account_payment_netting/models/account_payment.py:97]: Propose this another module for mixing invoice and refund to payment https://github.com/OCA/account-invoicing/pull/555

**Comment by kittiu** [comment]
Can this one be merged?

**Review by rousseldenis — APPROVED** [review]
Code review. LGTM

**Review by nikul-serpentcs — APPROVED** [review]
Code Review LGTM :+1: 

@yostashiro Can you please update your review?

**Comment by kittiu** [comment]
@yostashiro help help :)

**Comment by kittiu** [comment]
Hello, we are porting this feature into V13. Would be great if this will be merged. Thanks!
cc @kongrattapong

File: account_payment_netting/models/account_payment.py
Comment (yostashiro) [inline account_payment_netting/models/account_payment.py]: Should it be `commercial_partner_id` instead of `partner_id`?

File: account_payment_netting/models/account_payment.py
Comment (yostashiro) [inline account_payment_netting/models/account_payment.py]: invoice**s**

**Review by yostashiro — CHANGES_REQUESTED** [review]
Apologies that my review is long overdue. I still haven't checked the details of the code but I'm fine with a merge once the the part I point out on using `commercial_partner_id` is updated.

File: account_payment_netting/models/account_payment.py
Comment (kittiu) [reply account_payment_netting/models/account_payment.py]: Yep, fixed.

File: account_payment_netting/models/account_payment.py
Comment (kittiu) [reply account_payment_netting/models/account_payment.py]: edited.

**Review by yostashiro — APPROVED** [review]

**Pull request merged**
